File: cleaner.py
import os
import re
from docx import Document

# Folder containing the Word file(s)
folder_path = r"C:\Users\Madhan\Write Final"

# Expanded Italian accented character replacement map
char_map = {
    "à": "a", "À": "A",
    "è": "e", "È": "E",
    "é": "e", "É": "E",
    "ì": "i", "Ì": "I",
    "ò": "o", "Ò": "O",
    "ù": "u", "Ù": "U",
    "—":", "
}

# Words that should remain untouched
exceptional_words = {"μ", "µ"}  # Add more as needed

def clean_text(text, location="", table_mode=False):
    # Skip if text is exactly an exceptional word
    if text.strip() in exceptional_words:
        return text
    
    changes = []
    
    # Replace non-breaking space
    if "\u00A0" in text:
        changes.append("non-breaking space")
        text = text.replace("\u00A0", " ")
    
    # Replace Italian accented characters
    for k, v in char_map.items():
        if k in text:
            changes.append(k)
            text = text.replace(k, v)
    
    # Remove all other non-ASCII except bullet (•) and comma (,)
    text, removed_chars = remove_unwanted_specials(text)
    if removed_chars:
        changes.extend(removed_chars)
    
    # Print changes
    if changes:
        print(f"{location}: {', '.join(sorted(set(changes)))}")
    
    return text

def remove_unwanted_specials(text):
    removed_chars = []
    new_text = ""
    for ch in text:
        # Keep ASCII chars, bullet • (U+2022), and comma
        if ch.isascii() or ch == "•" or ch == ",":
            new_text += ch
        else:
            removed_chars.append(ch)
    return new_text, removed_chars

# Process all .docx files
for file_name in os.listdir(folder_path):
    if file_name.lower().endswith(".docx"):
        file_path = os.path.join(folder_path, file_name)
        print(f"Processing file: {file_name}")
        
        doc = Document(file_path)
        
        # Clean paragraphs
        for i, para in enumerate(doc.paragraphs):
            para.text = clean_text(para.text, f"Paragraph {i+1}")
        
        # Clean tables
        for ti, table in enumerate(doc.tables):
            for ri, row in enumerate(table.rows):
                for ci, cell in enumerate(row.cells):
                    cell.text = clean_text(cell.text, f"Table {ti+1} Row {ri+1} Col {ci+1}", table_mode=True)
        
        doc.save(file_path)
        print(f"Overwritten: {file_path}\n{'-'*50}")

print("All files processed.")

File: cleaner.js
import fs from 'fs';
import path from 'path';
import { Document, Packer, Paragraph, Table, TextRun } from 'docx'; // For reading, limited support
import readdirp from 'readdirp';

// Map of Italian accented characters to replacements
const charMap = {
  'à': 'a', 'À': 'A',
  'è': 'e', 'È': 'E',
  'é': 'e', 'É': 'E',
  'ì': 'i', 'Ì': 'I',
  'ò': 'o', 'Ò': 'O',
  'ù': 'u', 'Ù': 'U',
  '—': ', '
};

const exceptionalWords = new Set(['μ', 'µ']);

// Replace unwanted characters except ASCII, bullet (•), and comma (,)
function removeUnwantedSpecials(text) {
  let removedChars = [];
  let newText = '';
  for (const ch of text) {
    if (ch.charCodeAt(0) <= 127 || ch === '•' || ch === ',') {
      newText += ch;
    } else {
      removedChars.push(ch);
    }
  }
  return { newText, removedChars };
}

function cleanText(text, location = '', tableMode = false) {
  if (exceptionalWords.has(text.trim())) {
    return text;
  }

  const changes = [];

  if (text.includes('\u00A0')) {
    changes.push('non-breaking space');
    text = text.replace(/\u00A0/g, ' ');
  }

  for (const [k, v] of Object.entries(charMap)) {
    if (text.includes(k)) {
      changes.push(k);
      text = text.split(k).join(v);
    }
  }

  const { newText, removedChars } = removeUnwantedSpecials(text);
  if (removedChars.length) {
    changes.push(...removedChars);
  }
  text = newText;

  if (changes.length) {
    console.log(`${location}: ${[...new Set(changes)].sort().join(', ')}`);
  }

  return text;
}

async function processDocxFile(filePath) {
  console.log(`Processing file: ${path.basename(filePath)}`);

  // NOTE: docx npm package has limited reading support — for full editing use another library or
  // convert to plain text, edit, then recreate the docx

  // For demo, we'll read file buffer and create a new doc with cleaned paragraphs (losing formatting)

  // You might want to extract paragraphs with a library like mammoth
  // Here's a simplified way using mammoth:

  import mammoth from 'mammoth';
  const { value: text } = await mammoth.extractRawText({ path: filePath });

  // Split text into paragraphs roughly
  let paragraphs = text.split('\n').filter(p => p.trim() !== '');

  // Clean paragraphs
  paragraphs = paragraphs.map((p, i) => cleanText(p, `Paragraph ${i + 1}`));

  // Create new doc with cleaned paragraphs (basic, no tables)
  const doc = new Document({
    sections: [{
      children: paragraphs.map(p => new Paragraph(p)),
    }],
  });

  const buffer = await Packer.toBuffer(doc);

  await fs.promises.writeFile(filePath, buffer);

  console.log(`Overwritten: ${filePath}\n${'-'.repeat(50)}`);
}

async function main() {
  const folderPath = "C:\\Users\\Madhan\\Write Final";

  const files = await fs.promises.readdir(folderPath);

  for (const fileName of files) {
    if (fileName.toLowerCase().endsWith('.docx')) {
      const filePath = path.join(folderPath, fileName);
      await processDocxFile(filePath);
    }
  }

  console.log("All files processed.");
}

main().catch(console.error);
